### PR TITLE
Feature/order

### PR DESCRIPTION
**Pull request by tolstiachok123**
Added order model, business-logic for it and tests for it, refactored project

### CODE REVIEW

File: src/test/resources/application-test.yml [hunk 1]
```diff
@@ -7,6 +7,14 @@ spring:
 
   jpa:
     show-sql: true
+    properties:
+      hibernate:
+        dialect: org.hibernate.dialect.H2Dialect
+
+  h2:
```
Comment (DmitrySay) [inline src/test/resources/application-test.yml]: remove this block

File: src/test/java/com/academia/andruhovich/library/util/OrderHelper.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import static com.academia.andruhovich.library.util.UserHelper.createExistingUser;
+
+public class OrderHelper {
+
+    public static Order createExistingOrder() {
+        Order order = new Order();
+        order.setId(ID);
+        order.setUser(createExistingUser());
+        order.setTotalPrice(PRICE);
+//        order.setStatus(OrderStatus.PAID);
```
Comment (DmitrySay) [inline src/test/java/com/academia/andruhovich/library/util/OrderHelper.java]: remove comment

File: src/main/resources/db/changelog/changelog_001.xml [hunk 3]
```diff
@@ -7,107 +7,325 @@
...
-            CREATE TABLE `user_role` (
-            `user_id` BIGINT NOT NULL,
-            `role_id` BIGINT NOT NULL,
-            CONSTRAINT `fk_user_id` FOREIGN KEY (`user_id`) REFERENCES `user` (`id`),
-            CONSTRAINT `fk_role_id2` FOREIGN KEY (`role_id`) REFERENCES `role` (`id`)
-            );
-        </sql>
-    </changeSet>
-
+			CREATE TABLE IF NOT EXISTS `author`
```
Comment (DmitrySay) [inline src/main/resources/db/changelog/changelog_001.xml]: Please make this file human-readable

File: src/test/java/com/academia/andruhovich/library/util/UserHelper.java [hunk 4]
```diff
@@ -38,8 +36,8 @@ public static User createExistingUser() {
         user.setPassword(ENCRYPTED_PASSWORD);
         user.setFirstName(FIRST_NAME);
         user.setLastName(LAST_NAME);
-        user.setCreatedAt(ZonedDateTime.now(ZoneId.of(MINSK)));
-        user.setUpdatedAt(ZonedDateTime.now(ZoneId.of(MINSK)));
+        user.setCreatedAt(NOW);
```
Comment (DmitrySay) [inline src/test/java/com/academia/andruhovich/library/util/UserHelper.java]: use currentDate() from DateHelper

File: src/main/java/com/academia/andruhovich/library/model/User.java [hunk 5]
```diff
@@ -23,25 +32,28 @@
...
-			inverseJoinColumns = {@JoinColumn(name = "role_id")})
-	private Set<Role> roles;
+    @OneToMany(fetch = FetchType.EAGER)
+    @JoinTable(name = "user_role", joinColumns = {@JoinColumn(name = "user_id")},
+            inverseJoinColumns = {@JoinColumn(name = "role_id")})
+    private Set<Role> roles = new HashSet<>();
 
-	@Column(name = "password")
-	private String password;
+//    @OneToMany(fetch = FetchType.EAGER)
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/model/User.java]: remove comments

File: src/main/java/com/academia/andruhovich/library/dto/BookDto.java [hunk 6]
```diff
@@ -14,6 +16,9 @@
 @Data
 @NoArgsConstructor
 @AllArgsConstructor
+@JsonIdentityInfo(
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/dto/BookDto.java]: Can you explain to me the reason to use @JsonIdentityInfo ? 

Your regex is unclear for me. What do you check here?  a-zа-я - is it Russian? what do you do if you need add a third language? write custom validator depending on the locale
@Pattern(regexp = "[A-ZА-Я][a-zа-я]*[-\\s]?([A-ZА-Я]?[a-zа-я]*[-\\s]?)*")

File: src/main/java/com/academia/andruhovich/library/model/Tag.java [hunk 7]
```diff
@@ -19,6 +22,9 @@
 @AllArgsConstructor
 @Entity
 @Table(name = "tags")
+@JsonIdentityInfo(
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/model/Tag.java]: Can you explain to me the reason to use @JsonIdentityInfo ?

File: src/main/java/com/academia/andruhovich/library/model/Book.java [hunk 8]
```diff
@@ -28,7 +26,9 @@
 @AllArgsConstructor
 @Entity
 @Table(name = "book")
-//@EntityListeners(AuditingEntityListener.class)
+@JsonIdentityInfo(
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/model/Book.java]: Can you explain to me the reason to use @JsonIdentityInfo ?

File: src/main/java/com/academia/andruhovich/library/exception/GlobalExceptionHandler.java [hunk 9]
```diff
@@ -1,78 +1,115 @@
...
-		ExceptionDto exceptionDto = ExceptionDto.builder()
-				.status(NOT_FOUND.value())
-				.error(NOT_FOUND.getReasonPhrase())
-				.message(exception.getMessage())
-				.time(ZonedDateTime.now())
-				.build();
-		return ResponseEntity.status(NOT_FOUND).body(exceptionDto);
-	}
+    @ExceptionHandler(value = BusyEmailException.class)
+    public ResponseEntity<Object> handle(BusyEmailException ex) {
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/exception/GlobalExceptionHandler.java]: rename BusyEmailException  DuplicatedEmailException and 
The appropriate status code for "Already Exists" would be '409 Conflict'.  Update log message   log.error("Caught BusyEmailException: {}

File: src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+                .collect(Collectors.toSet());
+    }
+
+    @Transactional
+    @Override
+    public OrderResponseDto getOrder(Long id) {
+        User user = userService.getCurrent();
+        Order order = orderRepository.getByIdAndUserId(id, user.getId()).orElseThrow(() ->
+                new ResourceNotFoundException(String.format(ORDER_NOT_FOUND, id)));
+        if (DRAFT.equals(order.getStatus())) {
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java]: copy past with ?

private Order isNeedRecalculate(Order order) {
        if (DRAFT.equals(order.getStatus())) {
            return recalculatePriceAndHistory(order);
        }
        return order;
    }

File: src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+                        .findFirst()
+                        .orElseThrow(() ->
+                                new ResourceNotFoundException(String.format(BOOK_NOT_FOUND, id)))
+                )
+                .collect(Collectors.toSet());
+
+        order.setHistory(jsonConverter.collectionToJsonString(synchronisedOrderContentWrappers));
+        order.setTotalPrice(calculateTotalPrice(synchronisedOrderContentWrappers, synchronisedBooks));
+
+        return orderRepository.save(order);
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java]: move orderRepository.save(order); out of this method return  just return order

File: src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+                .orElseThrow(() ->
+                        new ResourceNotFoundException(String.format(BOOK_NOT_FOUND, book.getId())));
+        Integer quantity = currentBookAndQuantity.getQuantity();
+        BigDecimal quantityBigDecimal = BigDecimal.valueOf(quantity);
+        return book.getPrice().multiply(quantityBigDecimal);
+    }
+
+    protected Order recalculatePriceAndHistory(Order order) {
+        if (order.getHistory().startsWith("\"")) {
+            order.setHistory(order.getHistory().replace("\\", ""));
```
Comment (DmitrySay) [inline src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java]: move to sub method +add test
           if (order.getHistory().startsWith("\"")) {
            order.setHistory(order.getHistory().replace("\\", ""));
            order.setHistory(order.getHistory().substring(1, order.getHistory().length() - 1));
        }

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            .configure(DeserializationFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE, false)
+            .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+
+    public static String convertToJsonString(final Object obj) throws IOException {
+        objectMapper.registerModule(new JavaTimeModule());
+        return objectMapper.writeValueAsString(obj);
+    }
+
+    public static AuthResponseDto getAuthResponseDto(String response) {
+        AuthResponseDto authResponseDto = new AuthResponseDto();
```
Comment (DmitrySay) [inline src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: remove useless = new AuthResponseDto()

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+import static com.academia.andruhovich.library.exception.ErrorMessages.CANNOT_READ_JSON_VALUE;
+import static com.academia.andruhovich.library.util.DateHelper.CURRENT_LOCATION;
+
+@Data
+public class JsonConvertHelper {
+
+    private static final String CONTENT = "content";
+
+    private final static ObjectMapper objectMapper = new ObjectMapper()
```
Comment (DmitrySay) [inline src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: look same as  @Bean     public ObjectMapper ?   autoware

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            authResponseDto = objectMapper.convertValue(responseBody, new TypeReference<>() {
+            });
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(String.format(CANNOT_READ_JSON_VALUE, response));
+        }
+        return authResponseDto;
+    }
+
+    public static AuthorDto getAuthorDto(String response) {
+        AuthorDto authorDto = new AuthorDto();
```
Comment (DmitrySay) [inline src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: remove new AuthorDto();

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+            authorDto = objectMapper.convertValue(responseBody, new TypeReference<>() {
+            });
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(String.format(CANNOT_READ_JSON_VALUE, response));
+        }
+        return authorDto;
+    }
+
+    public static List<AuthorDto> getAuthorDtos(String response) {
+        List<AuthorDto> authors = new ArrayList<>();
```
Comment (DmitrySay) [inline src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: remove = new ArrayList<>(); do it below too

File: src/test/java/com/academia/andruhovich/library/controller/AuthorControllerTest.java [hunk 17]
```diff
@@ -55,62 +51,78 @@
...
+        String response = mockMvc.perform(MockMvcRequestBuilders
+                        .get("/api/authors")
+                        .accept(APPLICATION_JSON))
                 .andDo(print())
                 .andExpect(status().isOk())
-                .andExpect(MockMvcResultMatchers.jsonPath("$").exists())
-                .andExpect(MockMvcResultMatchers.jsonPath("$[*].id").isNotEmpty());
+                .andReturn().getResponse().getContentAsString();
+
+        List<AuthorDto> authors = getAuthorDtos(response);
```
Comment (DmitrySay) [inline src/test/java/com/academia/andruhovich/library/controller/AuthorControllerTest.java]: Have you tried to deserialize directly on List<AuthorDto>?

File: src/test/resources/application-test.yml (hunk 1 above)
Comment (tolstiachok123) [reply src/test/resources/application-test.yml]: Removed "h2:" block

File: src/test/java/com/academia/andruhovich/library/util/OrderHelper.java (hunk 2 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/util/OrderHelper.java]: removed comment

File: src/test/java/com/academia/andruhovich/library/util/UserHelper.java (hunk 4 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/util/UserHelper.java]: Fixed it

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java (hunk 14 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: Should i change this field on NOTstatic and all methods corse of @Autowired dont work with static?
image

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java (hunk 13 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: Removed it

File: src/main/java/com/academia/andruhovich/library/model/User.java (hunk 5 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/model/User.java]: Removed it

File: src/main/resources/db/changelog/changelog_001.xml (hunk 3 above)
Comment (tolstiachok123) [reply src/main/resources/db/changelog/changelog_001.xml]: Fixed it

File: src/main/java/com/academia/andruhovich/library/dto/BookDto.java (hunk 6 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/dto/BookDto.java]: Removed it

File: src/main/java/com/academia/andruhovich/library/model/Tag.java (hunk 7 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/model/Tag.java]: Removed it

File: src/main/java/com/academia/andruhovich/library/model/Book.java (hunk 8 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/model/Book.java]: Removed it

File: src/main/java/com/academia/andruhovich/library/exception/GlobalExceptionHandler.java (hunk 9 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/exception/GlobalExceptionHandler.java]: Fixed it

File: src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java (hunk 12 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java]: Refactored it and added test

File: src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java (hunk 10 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java]: Refactored it

File: src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java (hunk 11 above)
Comment (tolstiachok123) [reply src/main/java/com/academia/andruhovich/library/service/impl/OrderServiceImpl.java]: Refactored it

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java (hunk 14 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: Deleted JsonConvertHelper class

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java (hunk 15 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: Deleted JsonConvertHelper class

File: src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java (hunk 16 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/util/JsonConvertHelper.java]: Deleted JsonConvertHelper class

File: src/test/java/com/academia/andruhovich/library/controller/AuthorControllerTest.java (hunk 17 above)
Comment (tolstiachok123) [reply src/test/java/com/academia/andruhovich/library/controller/AuthorControllerTest.java]: Refactored it

**Pull request merged**